kernel/preemption/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+        apic_id,
+        preemption_was_enabled,
+    };
+
+    if preemption_was_enabled {
+        // When transitioning from preemption being enabled to disabled,
+        // we must disable the local APIC timer used for preemptive task switching.
+        apic::get_my_apic()
+            .expect("BUG: hold_preemption() couldn't get local APIC")
+            .write()
+            .enable_lvt_timer(false);
+    } else if prev_val == u8::MAX {
+        // Overflow occurred and the counter value wrapped around, which is a bug.
+        panic!("BUG: Overflow occurred in the preemption counter for CPU {}", apic_id);
+    }
+
+    guard
+}
+
+
+/// A guard type that ensures preemption is disabled as long as it is held.
+/// 
+/// Call [`hold_preemption()`] to obtain a `PreemptionGuard`.
+/// 
+/// Preemption *may* be re-enabled when this guard is dropped,
+/// but not necessarily so, because other tasks on this CPU 
+/// may also have acquired a `PreemptionGuard`.
+pub struct PreemptionGuard {
+    /// The APIC ID of the CPU on which preemption was held.
+    /// This is not necessary, it's just used for sanity checks or debugging.
+    apic_id: u8,
+    /// Whether preemption was enabled when this guard was created.
+    preemption_was_enabled: bool,
+}
+
+impl PreemptionGuard {
+    /// Returns `true` if preemption was originally enabled
+    /// when this guard was created and disabled them.
+    pub fn preemption_was_enabled(&self) -> bool {
+        self.preemption_was_enabled
+    }
+}
+
+impl Drop for PreemptionGuard {
+    fn drop(&mut self) {
+        let apic_id = apic::get_my_apic_id();
+        assert!(
+            self.apic_id == apic_id,
+            "PreemptionGuard::drop(): BUG: APIC IDs did not match! \
+            This indicates an unexpected task migration across CPUs."
+        );
+
+        let prev_val = PREEMPTION_COUNT[apic_id as usize].fetch_sub(1, Ordering::Relaxed);
+        if prev_val == 1 {
+            // If the previous counter value was 1, that means the current value is 1,
+            // which indicates we are transitioning from preemption disabled to enabled on this CPU.
+            // Thus, we re-enable the local APIC timer used for preemptive task switching.
+            apic::get_my_apic()
+                .expect("BUG: PreemptionGuard::drop() couldn't get local APIC")
+                .write()
+                .enable_lvt_timer(true);
+        } else if prev_val == 0 {
+            // Underflow occurred and the counter value wrapped around, which is a bug.
+            panic!("BUG: Underflow occurred in the preemption counter for CPU {}", apic_id);
+        }
+    }
+}
+
+
+/// Returns `true` if preemption is currently enabled on this CPU.
+/// 
+/// Note that this value can't be used as a lock indicator or property,
+/// as it is just a snapshot offers no guarantee that preemption will be enabled
+/// or disabled on the next attempt to enable/disable it
+/// or acquire a preemption-safe lock.
```

Comment:
It actually can be used as a lock indicator if it returns false. Only the current processor can reenable preemption, and since preemption is disabled, only the caller is running on the current processor.

Reply:
Yeah, i should change the docs. I actually started using it as such at the top of `schedule()` as an indicator, as seen in the description box of this PR.

Reply:
hmm, actually on second thought, this function alone cannot be used as an indicator of anything. I thought this comment was referring to `PreemptionGuard::preemption_was_enabled()`, which can be useful.

Reply:
Regardless, I've update the docs